File: .framework-tests/detox-react-native/tests/login.test.ts
import { device, element, by, expect as detoxExpect } from 'detox';

const credentials = {
  validUser: 'standard_user',
  lockedUser: 'locked_out_user',
  password: 'secret_sauce',
};

describe('SauceDemo Login', () => {
  beforeAll(async () => {
    await device.launchApp();
  });

  beforeEach(async () => {
    await device.reloadReactNative();
  });

  /**
   * Valid credentials navigate to the inventory screen.
   * 1. Launch the app and land on the login screen
   * 2. Enter username "standard_user" into the username field
   * 3. Enter password "secret_sauce" into the password field
   * 4. Tap the Login button
   * 5. Check: inventory screen is visible
   */
  // @tc:33110
  it('valid credentials navigate to inventory screen', async () => {
    await element(by.id('username')).typeText(credentials.validUser);
    await element(by.id('password')).typeText(credentials.password);
    await element(by.id('login-button')).tap();
    await detoxExpect(element(by.id('inventory-screen'))).toBeVisible();
  });

  /**
   * Locked-out user sees an error message.
   * 1. Launch the app and land on the login screen
   * 2. Enter username "locked_out_user" into the username field
   * 3. Enter password "secret_sauce" into the password field
   * 4. Tap the Login button
   * 5. Check: error message contains "locked out"
   */
  // @tc:33111
  it('locked out user sees error message', async () => {
    await element(by.id('username')).typeText(credentials.lockedUser);
    await element(by.id('password')).typeText(credentials.password);
    await element(by.id('login-button')).tap();
    await detoxExpect(element(by.id('error-message'))).toBeVisible();
    await detoxExpect(element(by.id('error-message'))).toHaveText(
      'Epic sadface: Sorry, this user has been locked out.'
    );
  });

  /**
   * Empty username shows a validation error.
   * 1. Launch the app and land on the login screen
   * 2. Leave the username field empty
   * 3. Enter password "secret_sauce" into the password field
   * 4. Tap the Login button
   * 5. Check: validation error "Username is required" is visible
   */
  // @tc:33112
  it('empty username shows validation error', async () => {
    await element(by.id('password')).typeText(credentials.password);
    await element(by.id('login-button')).tap();
    await detoxExpect(element(by.id('error-message'))).toHaveText(
      'Epic sadface: Username is required'
    );
  });

  /**
   * Logout returns the user to the login screen.
   * 1. Log in as standard_user
   * 2. Open the navigation menu
   * 3. Tap "Logout"
   * 4. Check: login screen is visible
   */
  // @tc:33113
  it('logout returns to login screen', async () => {
    await element(by.id('username')).typeText(credentials.validUser);
    await element(by.id('password')).typeText(credentials.password);
    await element(by.id('login-button')).tap();
    await detoxExpect(element(by.id('inventory-screen'))).toBeVisible();
    await element(by.id('react-burger-menu-btn')).tap();
    await element(by.id('logout_sidebar_link')).tap();
    await detoxExpect(element(by.id('login-button'))).toBeVisible();
  });
});

File: .framework-tests/flutter-dart/tests/login_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:saucedemo_app/main.dart';

void main() {
  group('SauceDemo Login', () {
    // @tc:33410
    // @smoke
    /// Valid credentials navigate to the inventory screen.
    ///
    /// 1. Launch the app on the login screen
    /// 2. Enter "standard_user" into the username field
    /// 3. Enter "secret_sauce" into the password field
    /// 4. Tap the Login button
    /// 5. Check: inventory screen is visible
    testWidgets('valid credentials navigate to inventory screen',
        (WidgetTester tester) async {
      await tester.pumpWidget(const SauceDemoApp());

      await tester.enterText(find.byKey(const Key('username')), 'standard_user');
      await tester.enterText(find.byKey(const Key('password')), 'secret_sauce');
      await tester.tap(find.byKey(const Key('login-button')));
      await tester.pumpAndSettle();

      expect(find.byKey(const Key('inventory-screen')), findsOneWidget);
    });

    // @tc:33411
    /// Locked-out user sees an error message.
    ///
    /// 1. Launch the app on the login screen
    /// 2. Enter "locked_out_user" into the username field
    /// 3. Enter "secret_sauce" into the password field
    /// 4. Tap the Login button
    /// 5. Check: error message "Sorry, this user has been locked out." is visible
    testWidgets('locked out user sees error message',
        (WidgetTester tester) async {
      await tester.pumpWidget(const SauceDemoApp());

      await tester.enterText(find.byKey(const Key('username')), 'locked_out_user');
      await tester.enterText(find.byKey(const Key('password')), 'secret_sauce');
      await tester.tap(find.byKey(const Key('login-button')));
      await tester.pumpAndSettle();

      expect(
        find.text('Epic sadface: Sorry, this user has been locked out.'),
        findsOneWidget,
      );
    });

    // @tc:33412
    /// Empty username shows a validation error.
    ///
    /// 1. Launch the app on the login screen
    /// 2. Leave the username field empty
    /// 3. Enter "secret_sauce" into the password field
    /// 4. Tap the Login button
    /// 5. Check: error message "Username is required" is visible
    testWidgets('empty username shows validation error',
        (WidgetTester tester) async {
      await tester.pumpWidget(const SauceDemoApp());

      await tester.enterText(find.byKey(const Key('password')), 'secret_sauce');
      await tester.tap(find.byKey(const Key('login-button')));
      await tester.pumpAndSettle();

      expect(find.text('Epic sadface: Username is required'), findsOneWidget);
    });

    // @tc:33413
    /// Logout returns the user to the login screen.
    ///
    /// 1. Log in as standard_user
    /// 2. Tap the navigation menu icon
    /// 3. Tap "Logout"
    /// 4. Check: login screen is visible
    testWidgets('logout returns to login screen', (WidgetTester tester) async {
      await tester.pumpWidget(const SauceDemoApp());

      await tester.enterText(find.byKey(const Key('username')), 'standard_user');
      await tester.enterText(find.byKey(const Key('password')), 'secret_sauce');
      await tester.tap(find.byKey(const Key('login-button')));
      await tester.pumpAndSettle();

      expect(find.byKey(const Key('inventory-screen')), findsOneWidget);
      await tester.tap(find.byKey(const Key('menu-button')));
      await tester.pumpAndSettle();
      await tester.tap(find.byKey(const Key('logout-link')));
      await tester.pumpAndSettle();

      expect(find.byKey(const Key('login-button')), findsOneWidget);
    });
  });
}
